server/lib/src/idm/identityverification.rs

```diff
@@ -0,0 +1,278 @@
+        if !identify_feature_is_available {
+            return Ok(Some(IdentifyUserResponse::IdentityVerificationUnavailable));
+        };
+        if ident.get_uuid().eq(&Some(*target)) {
+            return Ok(Some(IdentifyUserResponse::IdentityVerificationAvailable));
+        };
+        Ok(None)
+    }
+
+    pub fn handle_identify_user_start(
+        &mut self,
+        IdentifyUserStartEvent { target, ident }: &IdentifyUserStartEvent,
+    ) -> Result<IdentifyUserResponse, OperationError> {
+        if let Some(res) = self.check_for_early_return_conditions(target, ident)? {
+            return Ok(res);
+        }
+        let response = if ident.get_uuid() < Some(*target) {
+            IdentifyUserResponse::WaitForCode
+        } else {
+            let shared_secret = self.get_self_totp_secret(target, ident)?;
+            let totp = self.compute_totp(shared_secret)?;
+            IdentifyUserResponse::ProvideCode {
+                step: TOTP_STEP as u32,
+                totp,
+            }
+        };
+        Ok(response)
+    }
+
+    pub fn handle_identify_user_display_code(
+        &mut self,
+        IdentifyUserDisplayCodeEvent { target, ident }: &IdentifyUserDisplayCodeEvent,
+    ) -> Result<IdentifyUserResponse, OperationError> {
+        if let Some(res) = self.check_for_early_return_conditions(target, ident)? {
+            return Ok(res);
+        }
+        let shared_secret = self.get_self_totp_secret(target, ident)?;
+        let totp = self.compute_totp(shared_secret)?;
+        Ok(IdentifyUserResponse::ProvideCode {
+            step: TOTP_STEP as u32,
+            totp,
+        })
+    }
+
+    pub fn handle_identify_user_submit_code(
+        &mut self,
+        IdentifyUserSubmitCodeEvent {
+            target,
+            ident,
+            code,
+        }: &IdentifyUserSubmitCodeEvent,
+    ) -> Result<IdentifyUserResponse, OperationError> {
+        if let Some(res) = self.check_for_early_return_conditions(target, ident)? {
+            return Ok(res);
+        }
+        let other_user_totp_secret = self.get_other_user_totp_secret(target, ident)?;
+        let other_user_totp = self.compute_totp(other_user_totp_secret)?;
+        if other_user_totp != *code {
+            return Ok(IdentifyUserResponse::CodeFailure);
+        }
+        // if we are the fist it means now it's time to go for ProvideCode, otherwise we just confirm that the code is correct
+        // (we know this for a fact as we have already checked that the code is correct)
+        let res = if ident.get_uuid() < Some(*target) {
+            let shared_secret = self.get_self_totp_secret(target, ident)?;
+            let totp = self.compute_totp(shared_secret)?;
+            IdentifyUserResponse::ProvideCode {
+                step: TOTP_STEP as u32,
+                totp,
+            }
+        } else {
+            IdentifyUserResponse::Success
+        };
+        Ok(res)
+    }
+}
```

Comment:
Tests? You probably should get into the habit of writing the tests first before you impl a wall of feature. :)

Reply:
I was trusting the api tests to handle everything, yet after thinking about it I realized they depend on too many things, so more specific tests could help